From 007ede2721a637ad5b6d49c6170bb206ecf13422 Mon Sep 17 00:00:00 2001
From: Tom White <tom@cloudera.com>
Date: Tue, 24 Apr 2012 15:09:59 -0700
Subject: [PATCH 085/795] MR1: MAPREDUCE-2450. Calls from running tasks to TaskTracker methods sometimes fail and incur a 60s timeout.

Reason: Performance
Author: Rajesh Balamohan
Ref: CDH-5206
(cherry picked from commit 2f21b4b25af667475a46ca751186f16f214cac23)
(cherry picked from commit c0fb4a5154607c3b0c292edae7b8ed12e6307b5b)
---
 src/mapred/org/apache/hadoop/mapred/Task.java |   20 ++++++++++++++++++++
 1 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapred/Task.java b/src/mapred/org/apache/hadoop/mapred/Task.java
index 518d742..b17df5c 100644
--- a/src/mapred/org/apache/hadoop/mapred/Task.java
+++ b/src/mapred/org/apache/hadoop/mapred/Task.java
@@ -545,6 +545,8 @@ public void initialize(JobConf job, JobID id,
     private Progress taskProgress;
     private JvmContext jvmContext;
     private Thread pingThread = null;
+    private boolean done = true;
+    private Object lock = new Object();
     /**
      * flag that indicates whether progress update needs to be sent to parent.
      * If true, it has been set. If false, it has been reset. 
@@ -633,6 +635,9 @@ public void run() {
       // get current flag value and reset it as well
       boolean sendProgress = resetProgressFlag();
       while (!taskDone.get()) {
+        synchronized (lock) {
+          done = false;
+        }
         try {
           boolean taskFound = true; // whether TT knows about this task
           // sleep for a bit
@@ -665,6 +670,7 @@ public void run() {
           // came back up), kill ourselves
           if (!taskFound) {
             LOG.warn("Parent died.  Exiting "+taskId);
+            resetDoneFlag();
             System.exit(66);
           }
 
@@ -677,10 +683,19 @@ public void run() {
           if (remainingRetries == 0) {
             ReflectionUtils.logThreadInfo(LOG, "Communication exception", 0);
             LOG.warn("Last retry, killing "+taskId);
+            resetDoneFlag();
             System.exit(65);
           }
         }
       }
+      //Notify that we are done with the work
+      resetDoneFlag();
+    }
+    void resetDoneFlag() {
+      synchronized (lock) {
+        done = true;
+        lock.notify();
+      }
     }
     public void startCommunicationThread() {
       if (pingThread == null) {
@@ -692,6 +707,11 @@ public void startCommunicationThread() {
     public void stopCommunicationThread() throws InterruptedException {
       // Updating resources specified in ResourceCalculatorPlugin
       if (pingThread != null) {
+        synchronized (lock) {
+          while (!done) {
+            lock.wait();
+          }
+        }
         pingThread.interrupt();
         pingThread.join();
       }
-- 
1.7.0.4

